pkg/utils/managedresources/registry.go

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package managedresources
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/runtime/serializer/json"
+)
+
+// Registry stores objects and their serialized form. It allows to compute a map of all registered objects that can be
+// used as part of a Secret's data which is referenced by a ManagedResource.
+type Registry struct {
+	lock         sync.Mutex
+	codec        runtime.Codec
+	nameToObject map[string]*object
+}
+
+type object struct {
+	obj           runtime.Object
+	serialization []byte
+}
+
+// NewRegistry returns a new registry for resources. The given scheme, codec, and serializer must know all the resource
+// types that will later be added to the registry.
+func NewRegistry(scheme *runtime.Scheme, codec serializer.CodecFactory, serializer *json.Serializer) *Registry {
+	var groupVersions []schema.GroupVersion
+	for k := range scheme.AllKnownTypes() {
+		groupVersions = append(groupVersions, k.GroupVersion())
+	}
+
+	return &Registry{
+		codec:        codec.CodecForVersions(serializer, serializer, schema.GroupVersions(groupVersions), schema.GroupVersions(groupVersions)),
+		nameToObject: make(map[string]*object),
+	}
+}
+
+// Add adds the given object the registry. It computes a filename based on its type, namespace, and name. It serializes
+// the object to YAML and stores both representations (object and serialization) in the registry.
+func (r *Registry) Add(obj runtime.Object) error {
+	r.lock.Lock()
+	defer r.lock.Unlock()
```

Comment:
Is locking really necessary for the registry? I probably won't/shouldn't be used concurrently in multiple goroutines, right?
If we have several registries per ongoing shoot reconciliation each locking and unlocking with high frequency, I would imagine that it _could_ have some performance impacts.

Reply:
Yeah, I wasn't even sure myself when I implemented it. I wanted to be on the safe side, but your comment confirms my gud feeling that it's most likely not necessary. I'll remove it, thanks!